R/tree-sequences.R

```diff
@@ -83,30 +111,20 @@ ts_load <- function(model, file = NULL,
     stop("Mutation rate must be given in order to be able to mutate the tree sequence",
          call. = FALSE)
 
-  if (!inherits(model, "slendr_model"))
-    stop("A compiled slendr model object must be provided", call. = FALSE)
-
-  if (is.null(file)) {
-    ts_files <- list.files(model$path, pattern = "_(msprime|slim)\\.trees$", full.names = TRUE)
-    if (length(ts_files) == 1)
-      file <- ts_files
-    else if (length(ts_files) > 1) {
-      stop("Multiple tree sequence files found in the model directory:\n",
-              paste(ts_files, collapse = "\n"),
-              "\n\nPlease use the `file = ` argument of `ts_load()` to specify ",
-              "which tree sequence should be loaded.", call. = FALSE)
-    } else
-      stop("No SLiM or msprime tree sequence file found in the model directory ",
-           model$path, call. = FALSE)
-  }
-
   # load the tree sequence, converting it to a SLiM tree sequence if necessary
   ts <- tskit$load(path.expand(file))
-  if (!is.null(ts$metadata$SLiM)) {
+
+  # this is hacky but it doesn't seem worth it parsing the raw provenance
+  # string that tskit seems to provide, so just regexing against the software
+  # is good enough for now
+  provenance <- ts$provenance(0L)$record
+  if (grepl("SLiM", provenance)) {
     ts <- pyslim$SlimTreeSequence(ts)
     backend <- "SLiM"
-  } else
+  } else if (grepl("msprime", provenance))
     backend <- "msprime"
+  else
+    stop("Only tree sequences produced by SLiM or msprime are supported", call. = FALSE)
```

Comment:
The right way to check if something is SLiM-produced is to check in top-level metadata? (i.e, check for `ts.metadata['SLiM']`) However, do you really need to check this at all? You don't want to stop someone else from using slendr on a valid tree sequence made some other way?

Reply:
Hmm, I don't think `ts.metadata["msprime"]` is a thing though? At least in my msprime runs that field is empty? I thought that "provenance" is exactly what should specify the source of the tree sequence (and checking the dictionary to see what the word "provenance" even means confirmed my assumption :D). If whatever produced a tree sequence doesn't populate the metadata field (as seems to be the case for msprime), then I can't know the source right?

Or is it that there are only two classes of tree sequences out there? SLiM and everything else? In that case, it would be enough to have SLiM-specific table munging code and the msprime-specific code could be used for msprime and everything else. Rather than assuming "everything else" will require dedicated code (on top of SLiM-specific and msprime-specific code).

I guess I don't know quite enough about all other possible sources of tree sequence data out there.

Reply:
Here's the situation: msprime tree sequences have just the basic stuff; SLiM tree sequences have extra metadata of a particular sort. There may be other sort out there, but none that slendr has to worry about, that I know of. So, I think you can have (a) code that applies to all tree sequences, and (b) code that does extra stuff for SLiM tree sequences.

But - what's the 'extra stuff' you do for SLiM?

Reply:
This is very helpful, thank you!

If msprime tree sequences are no different from any other tskit tree sequence (either constructed manually, or obtained through something like tsinfer, etc.) then yes, anything that has something to do with msprime can be used for "generic" tree sequences, no problem there.

Main extra stuff that happens for SLiM tree sequences is:

- If those are generated from slendr models, that "extra stuff" would be:
    - Assigning symbolic names to individuals (and nodes) so that they can be referred to by users in simplification or computing statistics, outputting VCF or EIGENSTRAT, directly by names (i.e. `EUR_42` based on a user-defined sampling table, instead of numerical node IDs).
        - Also maintaining those names through simplification (because tskit renames node/individual IDS, I'm using `pedigree_id` from SLiM through join operations to know what got dropped and what is kept in the tree sequence)

    - Times of nodes and mutations, etc are optionally transformed from tskit units of backward time in generations into the time units used by the user to define models ("years before present", "years in the future from today", arbitrary time windows basically).
 
- If a SLiM tree sequence is spatial (either from slendr or pure SLiM), then locations are transformed either to the appropriate CRS used by the model in the [sf](https://r-spatial.github.io/sf/articles/sf1.html) format.

Taking care of the above is what most of the `!is.null(model)` conditions do -- non-slendr tree sequences don't have the additional metadata created by slendr itself (sampling tables with names of individuals to be recorded, coordinate reference systems, custom time units), but all the other SLiM tree sequence table munging is the same which is why I was able use the same code for this and only switch off the optional slendr-specific annotation code when needed.

Some of the points above (except the spatial stuff) applies also to slendr's msprime tree sequences (named individuals, custom times of nodes). There is currently exactly one function which has `msprime` in its name (`get_msprime_table_data()`) and that takes care of all of this. 

So even plugging in a "generic" tree sequence object should work without any changes to the internals of that function.

Is there a large-ish non-SLiM and non-msprime tree sequence file somewhere in tskit testing suite that I could borrow to test this on a non-simulated tree sequence?

---


**One thing I have to make very clear: the what I call 'raw' tree sequence tables (edges, nodes, individuals, mutations) are never touched by this R code. They are always accessible <u>in their original form</u>  via `ts_table(ts_object, "<edges|nodes|individuals|mutations>")`! None of the code munging you saw modifies those or adds additional information.** Most of the `get_slim_table_data` and `get_msprime_table_data` produces a more richly annotated tables useful for data analysis (with readable symbolic names of nodes of sampled individuals, model-specific times, geographic coordinates, etc). Those are available under `ts_nodes()` and `ts_edges()` in a rich annotated form (basically what can be directly used for visualization or statistics).